README.md
# servant-crud

# Build

```
stack build
```

# Run

```
stack exec servant-crud-exe
```

# API

```
# get user list
curl -XGET 127.0.0.1:8081/users

# add new user

# get user by ID
curl -XGET 127.0.0.1:8081/users/1

# update user

# delete user
curl -XDELETE 127.0.0.1:8081/users/1
```

# Build with Docker and create executable Docker image (WIP)

```
make all
```

# Run with Docker (WIP)

```
docker-compose up
```


ChangeLog.md
# Changelog for servant-crud

## Unreleased changes


src/SampleServer.hs
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE DeriveGeneric         #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE RankNTypes            #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE TypeOperators         #-}

module SampleServer
  ( app
  ) where

import           Control.Concurrent
import           Control.Concurrent.STM
import           Control.Monad.IO.Class ( liftIO )
import           Data.Aeson
import           Data.Time              ( UTCTime )
import           GHC.Generics           ( Generic )
import           Servant

import           ServantCrud            ( APIFor )

data User =
  User
    { userFirstName             :: String
    , userLastName              :: String
    , userEmail                 :: String
    , userRegistrationTimeStamp :: UTCTime
    }
  deriving (Eq, Show, Generic)

instance FromJSON User

instance ToJSON User

userId1 = 1 :: Int

user1 =
  User
    "Isaac"
    "Newton"
    "isaac@newton.co.uk"
    (read "1683-03-01 00:00:00" :: UTCTime)

userId2 = 2 :: Int

user2 =
  User
    "Albert"
    "Einstein"
    "albert@mc2.org"
    (read "1905-12-01 00:00:01" :: UTCTime)

users = [(userId1, user1), (userId2, user2)]

type UserAPI = "users" :> APIFor User Int

userAPI :: Proxy UserAPI
userAPI = Proxy

type UserId = Int

type UserRecord = (UserId, User)

type Pool = TVar [UserRecord]

initUserPool :: IO Pool
initUserPool = atomically $ newTVar users

getUsers :: Pool -> IO [UserRecord]
getUsers = readTVarIO

newUser :: Pool -> User -> IO UserId
newUser pool user = do
  users <- getUsers pool
  let newId = (maximum . map fst) users + 1
      updatedUsers = users ++ [(newId, user)]
  atomically $ writeTVar pool updatedUsers
  return newId

-- FIXME: should not throw runtime exception even if user ID not found
getUser :: Pool -> UserId -> IO User
getUser pool id' = do
  users <- getUsers pool
  return $ head [user | (userId, user) <- users, userId == id']

-- FIXME: should not throw runtime exception even if user ID not found
updateUser :: Pool -> UserId -> User -> IO ()
updateUser pool id' newUser = do
  users <- getUsers pool
    -- just check the existence of id'
  let _ = head [user | (userId, user) <- users, userId == id']
      replaceUser (userId, user) userId' newUser =
        case userId of
          userId' -> (userId, newUser)
          _       -> (userId, user)
      updatedUsers = map (\(i, u) -> replaceUser (i, u) id' newUser) users
  atomically $ writeTVar pool updatedUsers

deleteUser :: Pool -> UserId -> IO ()
deleteUser pool id' = do
  users <- getUsers pool
  let updatedUsers = [x | x <- users, fst x /= id']
  atomically $ writeTVar pool updatedUsers

userServer :: Pool -> Server UserAPI
userServer pool = getEntities pool :<|> newEntity pool :<|> operations pool
  where
    getEntities = liftIO . getUsers
    newEntity pool user = liftIO $ newUser pool user >> return NoContent
    operations pool id' =
      getEntity pool id' :<|> updateEntity pool id' :<|> deleteEntity pool id'
    getEntity pool id' = liftIO $ getUser pool id'
    updateEntity pool id' user =
      liftIO $ updateUser pool id' user >> return NoContent
    deleteEntity pool id' = liftIO $ deleteUser pool id' >> return NoContent

server :: IO (Server UserAPI)
server = userServer <$> initUserPool

app :: IO Application
app = serve userAPI <$> server


src/PersistentUtil.hs
{-# LANGUAGE GADTs #-}

module PersistentUtil
  (
  ) where

import           Conduit                 ( MonadUnliftIO )
import           Control.Monad.Logger    ( runStderrLoggingT )
import           Data.Pool               ( Pool )
import           Data.String.Conversions ( cs )
import           Database.Persist
import           Database.Persist.Sql
import           Database.Persist.Sqlite
import           Database.Persist.TH

type NumConnection = Int

mkSqlitePool :: FilePath -> NumConnection -> IO ConnectionPool
mkSqlitePool fp n = runStderrLoggingT $ createSqlitePool (cs fp) n

{-
 - CRUD functions
 -}
selectList' ::
     ( PersistQueryRead backend
     , IsPersistBackend backend
     , MonadUnliftIO m
     , PersistEntity record
     , PersistEntityBackend record ~ BaseBackend backend
     , BaseBackend backend ~ SqlBackend
     )
  => Data.Pool.Pool backend
  -> m [Entity record]
selectList' = runSqlPool $ selectList [] []

get' ::
     ( PersistStoreRead backend
     , IsPersistBackend backend
     , MonadUnliftIO m
     , PersistEntity record
     , PersistEntityBackend record ~ BaseBackend backend
     , BaseBackend backend ~ SqlBackend
     )
  => Key record
  -> Data.Pool.Pool backend
  -> m (Maybe record)
get' = runSqlPool . get

insert' ::
     ( PersistStoreWrite backend
     , IsPersistBackend backend
     , MonadUnliftIO m
     , PersistEntity record
     , PersistEntityBackend record ~ BaseBackend backend
     , BaseBackend backend ~ SqlBackend
     )
  => record
  -> Data.Pool.Pool backend
  -> m (Key record)
insert' = runSqlPool . insert

update' ::
     ( PersistStoreWrite backend
     , IsPersistBackend backend
     , MonadUnliftIO m
     , PersistEntity record
     , PersistEntityBackend record ~ BaseBackend backend
     , BaseBackend backend ~ SqlBackend
     )
  => Key record
  -> [Update record]
  -> Data.Pool.Pool backend
  -> m record
update' k rs = runSqlPool $ updateGet k rs

delete' ::
     ( PersistStoreWrite backend
     , IsPersistBackend backend
     , MonadUnliftIO m
     , PersistEntity record
     , PersistEntityBackend record ~ BaseBackend backend
     , BaseBackend backend ~ SqlBackend
     )
  => Key record
  -> Data.Pool.Pool backend
  -> m ()
delete' = runSqlPool . delete


src/SampleServerDB.hs
{-# LANGUAGE DeriveGeneric              #-}
{-# LANGUAGE EmptyDataDecls             #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE FlexibleInstances          #-}
{-# LANGUAGE GADTs                      #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE QuasiQuotes                #-}
{-# LANGUAGE StandaloneDeriving         #-}
{-# LANGUAGE TemplateHaskell            #-}
{-# LANGUAGE TypeFamilies               #-}

module SampleServerDB
  ( server
  ) where

import           Control.Monad.Logger    ( runStderrLoggingT )
import           Data.Aeson              ( FromJSON, ToJSON )
import           Data.String.Conversions ( cs )
import           Data.Time               ( UTCTime )
import           Database.Persist
import           Database.Persist.Sql
import           Database.Persist.Sqlite
import           Database.Persist.TH
import           GHC.Generics            ( Generic )
import           ServantCrud             ( APIFor )

share
  [mkPersist sqlSettings, mkMigrate "migrateAll"]
  [persistLowerCase|
User
  firstName String
  lastName String
  email String
  registrationTimeStamp UTCTime default=CURRENT_TIMESTAMP
  deriving Read Eq Generic Show
|]

instance FromJSON User

instance ToJSON User

user1 =
  User
    "Isaac"
    "Newton"
    "isaac@newton.co.uk"
    (read "1683-03-01 00:00:00" :: UTCTime)

user2 =
  User
    "Albert"
    "Einstein"
    "albert@mc2.org"
    (read "1905-12-01 00:00:01" :: UTCTime)

mkPool :: FilePath -> IO ConnectionPool
mkPool filePath = runStderrLoggingT $ createSqlitePool (cs filePath) 10

initialize :: ConnectionPool -> IO ()
initialize = runSqlPool (runMigration migrateAll)

server :: IO ()
server = do
  let _database_ = "SampleServerDB-dev"
  pool <- mkPool _database_
  initialize pool


src/SampleServerIMDB.hs
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE DeriveGeneric         #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE RankNTypes            #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE TypeOperators         #-}

module SampleServerIMDB
  ( app
  ) where

import           Control.Concurrent
import           Control.Concurrent.STM
import           Control.Monad.IO.Class ( liftIO )
import           Data.Aeson
import           Data.Time              ( UTCTime )
import           GHC.Generics           ( Generic )
import           Servant

import           ServantCrud            ( APIFor )

data User =
  User
    { userFirstName             :: String
    , userLastName              :: String
    , userEmail                 :: String
    , userRegistrationTimeStamp :: UTCTime
    }
  deriving (Eq, Show, Generic)

instance FromJSON User

instance ToJSON User

userId1 = 1 :: Int

user1 =
  User
    "Isaac"
    "Newton"
    "isaac@newton.co.uk"
    (read "1683-03-01 00:00:00" :: UTCTime)

userId2 = 2 :: Int

user2 =
  User
    "Albert"
    "Einstein"
    "albert@mc2.org"
    (read "1905-12-01 00:00:01" :: UTCTime)

users = [(userId1, user1), (userId2, user2)]

type UserAPI = "users" :> APIFor User Int

userAPI :: Proxy UserAPI
userAPI = Proxy

type UserId = Int

type UserRecord = (UserId, User)

type Pool = TVar [UserRecord]

initUserPool :: IO Pool
initUserPool = atomically $ newTVar users

getUsers :: Pool -> IO [UserRecord]
getUsers = readTVarIO

newUser :: Pool -> User -> IO UserId
newUser pool user = do
  users <- getUsers pool
  let newId = (maximum . map fst) users + 1
      updatedUsers = users ++ [(newId, user)]
  atomically $ writeTVar pool updatedUsers
  return newId

-- FIXME: should not throw runtime exception even if user ID not found
getUser :: Pool -> UserId -> IO User
getUser pool id' = do
  users <- getUsers pool
  return $ head [user | (userId, user) <- users, userId == id']

-- FIXME: should not throw runtime exception even if user ID not found
updateUser :: Pool -> UserId -> User -> IO ()
updateUser pool id' newUser = do
  users <- getUsers pool
    -- just check the existence of id'
  let _ = head [user | (userId, user) <- users, userId == id']
      replaceUser (userId, user) userId' newUser =
        case userId of
          userId' -> (userId, newUser)
          _       -> (userId, user)
      updatedUsers = map (\(i, u) -> replaceUser (i, u) id' newUser) users
  atomically $ writeTVar pool updatedUsers

deleteUser :: Pool -> UserId -> IO ()
deleteUser pool id' = do
  users <- getUsers pool
  let updatedUsers = [x | x <- users, fst x /= id']
  atomically $ writeTVar pool updatedUsers

userServer :: Pool -> Server UserAPI
userServer pool = getEntities pool :<|> newEntity pool :<|> operations pool
  where
    getEntities = liftIO . getUsers
    newEntity pool user = liftIO $ newUser pool user >> return NoContent
    operations pool id' =
      getEntity pool id' :<|> updateEntity pool id' :<|> deleteEntity pool id'
    getEntity pool id' = liftIO $ getUser pool id'
    updateEntity pool id' user =
      liftIO $ updateUser pool id' user >> return NoContent
    deleteEntity pool id' = liftIO $ deleteUser pool id' >> return NoContent

server :: IO (Server UserAPI)
server = userServer <$> initUserPool

app :: IO Application
app = serve userAPI <$> server
